libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigationNotification.kt

```diff
@@ -0,0 +1,380 @@
+        * */
+
+        // *** Converted code
+        generateArrivalTime(routeProgress, time)?.let { formattedTime ->
+            updateViewsWithArrival(formattedTime)
+            routeProgress.currentLegProgress().upComingStep()?.let { step ->
+                routeProgress.currentLegProgress().upComingStep()
+                updateManeuverImage(step)
+            } ?: routeProgress.currentLegProgress().currentStep()
+        }
+    }
+
+    // Package private (no modifier) for testing purposes
+    fun retrieveInstructionText(): String? {
+        return instructionText
+    }
+
+    // Package private (no modifier) for testing purposes
+    fun retrieveCurrentManeuverId(): Int {
+        return currentManeuverId
+    }
+
+    private fun initialize(applicationContext: Context, mapboxNavigation: MapboxNavigation) {
+        this.mapboxNavigation = mapboxNavigation
+        etaFormat = applicationContext.getString(R.string.eta_format)
+        initializeDistanceFormatter(applicationContext, mapboxNavigation)
+        applicationContext.getSystemService(Context.NOTIFICATION_SERVICE)?.let { notificationService ->
+            notificationManager = notificationService as NotificationManager
+        }
+        isTwentyFourHourFormat = DateFormat.is24HourFormat(applicationContext)
+
+        pendingOpenIntent = createPendingOpenIntent(applicationContext)
+        pendingCloseIntent = createPendingCloseIntent(applicationContext)
+
+        registerReceiver(applicationContext)
+        createNotificationChannel(applicationContext)
+        if (notification == null) {
+            notification = buildNotification(applicationContext)
+        }
+    }
+
+    private fun initializeDistanceFormatter(applicationContext: Context, mapboxNavigation: MapboxNavigation) {
+        val routeOptions = mapboxNavigation.route.routeOptions()
+        val localeUtils = LocaleUtils()
+        var language: String = localeUtils.inferDeviceLanguage(applicationContext)
+        var unitType: String = localeUtils.getUnitTypeForDeviceLocale(applicationContext)
+        routeOptions?.let { options ->
+            language = options.language() ?: "en"
+            unitType = options.voiceUnits() ?: IMPERIAL
+        }
+
+        val mapboxNavigationOptions = mapboxNavigation.options()
+        val roundingIncrement = mapboxNavigationOptions.roundingIncrement()
+
+        distanceFormatter = DistanceFormatter(applicationContext, language, unitType, roundingIncrement)
+    }
+
+    private fun createNotificationChannel(applicationContext: Context) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val notificationChannel = NotificationChannel(
+                    NavigationConstants.NAVIGATION_NOTIFICATION_CHANNEL, applicationContext.getString(R.string.channel_name),
+                NotificationManager.IMPORTANCE_LOW
+            )
+            notificationManager?.createNotificationChannel(notificationChannel)
```

Comment:
Do we need a `let` call here? What happens if `notificationManager` is null?

Reply:
If notificationManager is null, no call is made. No need for let.